### DISCUSSION TITLE
[turborepo] turbo-ignore always builds all projects for new branches on Vercel

### DISCUSSION THREAD

**Discussion by kieranm** [vercel/turborepo#7455] [Ideas]
### What version of Turborepo are you using?

turbo-ignore@1.10.16

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Mac

### Describe the Bug

We have a monorepo with multiple websites, and for each Vercel app we have set the ignored build step to `npx turbo-ignore`. We have the Git integration enabled for preview builds.

When we open a PR, we expect that only changed apps will be built, to reduce unnecessary builds.

However, in Vercel I see the following output:

```
Running "npx turbo-ignore"
--
22:51:51.448 | npm WARN exec The following package was not found and will be installed: turbo-ignore@1.10.16
22:51:51.767 | ≫   Using Turborepo to determine if this project is affected by the commit...
22:51:51.768 |  
22:51:51.772 | ≫   Inferred "@daybridge/daybridge.design" as workspace from "package.json"
22:51:51.772 | ≫   Using "build" as the task from the arguments
22:51:51.773 | ≫   No previous deployments found for "@daybridge/daybridge.design" on branch "day-2084-item-form-does-not-appear-for-items-ending-at-midnight-on"
22:51:51.773 | ✓ Proceeding with deployment
```

The `@daybridge/daybridge.design` app did not have any changed files, so it should have been ignored. But the message suggests that all web apps are being built because they haven't been deployed before on this branch.

### Expected Behavior

I'm not sure if I'm doing something wrong, or if there is some context I'm missing, but it's important that apps that have not been changed don't deploy unnecessarily, so I would expect turbo-ignore to be skipping these even if they haven't been deployed before for a specific branch.

### To Reproduce

- Monorepo with one or more web projects
- Each project added as a separate app to Vercel, with preview builds enabled
- Each web project has its ignored build step set to `npx turbo-ignore`
- Make a change impacting only one of the web projects and none of the others
- Open PR in Github

**Comment by gsoltis** [comment]
Can you try setting `--fallback` to handle scenarios when it's the first deploy of a particular branch?

See [Usage](https://www.npmjs.com/package/turbo-ignore#usage)

**Comment by kieranm** [comment]
Thanks! I can definitely try setting this, but the guidance on Vercel's docs says its only necessary to use `npx turbo-ignore` without any additional flags. This used to work just fine. So I would like to understand why this has started happening, and potentially whether its a bug.

**Comment by kieranm** [comment]
In fact, in Vercel this field is actually disabled so it's not possible to change it without bailing out and setting a custom command. Despite having this set, all my apps are currently being built even if there are no changes.

<img width="955" alt="Screenshot 2023-10-30 at 10 19 44" src="https://github.com/vercel/turbo/assets/232283/d157584f-3b9d-4098-a6d8-a71b65533770">

**Comment by gsoltis** [comment]
Is this only for the first commit of a new branch, or is this happening on every subsequent commit as well?

**Comment by tknickman** [comment]
Hey @kieranm! While strange, this is intended behavior. Here is why we decided to make this change. 

Originally, when `turbo-ignore` was experimental, we would always compare `HEAD` against `HEAD^1` (the _parent_ commit) to determine changes. This works _most_ of the time, and is still the default when **not** deploying on Vercel. However, there are some corner cases that make this a dangerous default. The most obvious one is when the branch you are deploying has multiple commits. 

Say you have a branch - `feat-1` with the following commits:
`ghi`: a `web` change
`def`: a `web` change
`abc`: a `docs` change

When deploying `feat-1`, if using `HEAD^1` as the comparison, _only_ `web` would be identified as having changes, and thus would be the only app deployed, despite `docs` also having changes in the branch. This is really frustrating and confusing, and requires manual intervention from the end-user (you would have to redeploy).

To fix this, we don't compare against `HEAD^1`, we compare against the _last commit that was deployed for the current branch_. This ensures that you're always checking to see if _anything_ from the last deployed commit up to and including your _current_ branch `HEAD` contains any relevant changes. 

Now, the issue you're experiencing comes in when the branch has never been deployed before. In this case, we don't have anything to compare against. So instead of falling back to `HEAD^1` which _could_ be frustratingly incorrect (pushing a new branch and having deploys skipped for no reason is not fun) as described above, we stay on the safe side and _always_ deploy. 

As @gsoltis pointed out, we do provide an escape hatch for this via the `--fallback` arg:
```
-f, --fallback <ref>       On Vercel, if no previously deployed SHA is available to compare against, fallback to comparing against the provided ref
```

This allows you to provide a comparison to use in the event that the previously deployed commit is not available (first push of branch etc.)


We're happy to entertain better options here if you have some suggestions as this is something that we fully admit is a bit confusing. But our goal is for `turbo-ignore` to never _incorrectly_ skip a deploy. This does mean that sometimes if we can't come to a conclusive result we opt for the safe approach of deploying just in case.

**Comment by kieranm** [comment]
Thanks for the explanation! The behaviour makes sense

If there are no previous deploys for a branch, is there a way of checking for changes by comparing with the last deploy of the branch you are merging **into**? Eg. if `feat-1` was being merged into `staging`, look at the last successfully deployed `staging` commit and then compare against that?

In any case, I would suggest that the copy in the screenshot on Vercel could change or be clarified, because it's currently a bit misleading if apps are going to be deployed for safety even if they haven't changed vs the base branch. The behaviour is sensible but just needs to be made clearer I think.

**Comment by tknickman** [comment]
Yes that would be the best solution here, and we're working on something internally to solve this that should be available soon. 

Right now, this is a limitation on Vercel due to the way the repo is cloned (to keep clones speedy, we only get details about the current branch, no others) so Vercel doesn't know anything about the base to compare against.

**Reply by octref** [reply]
Do you have an ETA for this feature? Currently a new PR to update README would build all of our 5 projects for preview.

**Comment by octref** [comment]
@tknickman I'm trying to use a custom script to figure out SHA of main branch and provide that to `--fallback`. However, because of Vercel's shallow clone on the build machine, I can't even find the SHA of `main`.

**Reply by octref** [reply]
I opened https://github.com/vercel/turbo/issues/7724

**Comment by valeriangalliat** [comment]
I've come up with a workaround in the meantime

- Create a SSH keypair e.g. `ssh-keygen -t ed25519 -f vercel`
- Add the public key as a read-only deploy key in your GitHub repo
- Add the private key in your project's env vars on Vercel e.g. `GITHUB_DEPLOY_KEY`

Then run the following script as part of `ignoreCommand`:

```sh
mkdir -p ~/.ssh
echo "$GITHUB_DEPLOY_KEY" > ~/.ssh/id_ed25519
chmod 0600 ~/.ssh/id_ed25519
git remote add origin git@github.com:your-org/your-repo
GIT_SSH_COMMAND='ssh -o StrictHostKeyChecking=no -i ~/.ssh/id_ed25519' git fetch origin main:refs/remotes/origin/main
npx turbo-ignore --fallback origin/main
```

There's a slight overhead to doing that, but most of the time there's not many extra refs to fetch between the shallow clone and `main`, and in our case with a monorepo that contains multiple Vercel projects, it's much better than having each new branch triggering a build in all the Vercel projects even if only touches one.

**Reply by Ghislain89** [reply]
@valeriangalliat how do you deal with main and production deployments? If the above scripts would run on main for a production deployment, it would compare ``main`` to `origin/main` and result in a 0 exit code, right?

**Comment by anthonyshew** [comment]
We've improved Turborepo on Vercel such that you no longer need to use this `turbo-ignore`. Instead, we encourage you to use the new setting to automatically ignore projects without changes: https://vercel.com/changelog/automatically-skip-unnecessary-deployments-in-monorepos-6UleK0qIbq8MByAtP6HP5U

**Reply by kaisermann** [reply]
Hey there 👋 Does using a custom ignore script prevent `skip unnecessary deployments` from running? My team has a monorepo with 3 apps (each app with a staging and production build). All of them use the `turbo-ignore` ignore script and have the "skip unnecessary deployments" turned on. We are still seeing at least one build on the first commit of a PR for each app/environment, even if they were unaffected.  Am I missing something 😅?

**Reply by Ghislain89** [reply]
I think that is because you're using turbo ignore. As Vercel does a shallow clone, it will not be able to determine if the app changed compared to previous deployment (there is no history). it will always deploy the first commit and for the next build, it should skip it again.

I think if you stop using Turbo ignore, you should get the expected behavior: do not deploy if unchanged. I'm not entirely sure what the skip unnecessary deployments does under the hood, but I expect it clones the full repo under the hood and uses that to determine the diff between your commit and the previous.

**Reply by anthonyshew** [reply]
When using the feature mentioned above, you will want to remove turbo-ignore. It's a more intelligent superset of turbo-ignore.

**Reply by kaisermann** [reply]
@anthonyshew I guess that, thanks. My question, however, is if having `turbo-ignore` or better, a custom ignore script, disables the `skip unnecessary deployments` feature.

I ask this because we use a custom ignore script to allow only certain builds in production. Roughly:

Production Builds:
- The deployment must be on a release branch (e.g., main).
- The latest commit message must match a specific release message (default: "Update versions").
- The turbo-ignore check must detect changes in the workspace.

Preview Builds:
- The deployment must be from a pull request (PR).
- The turbo-ignore check must detect changes.
- If no PR exists or turbo-ignore finds no changes, the build is ignored.

**Reply by ocelot2123** [reply]
> When using the feature mentioned above, you will want to remove [turbo-ignore](https://linear.app/vercel/issue/TURBO-ignore). It's a more intelligent superset of [turbo-ignore](https://linear.app/vercel/issue/TURBO-ignore).

Is the feature of turbo-ignore `[vercel only <workspace>]` also available in Vercel's custom feature?
